controllers/activegate/capability/capability.go

```diff
@@ -30,169 +40,235 @@ type Configuration struct {
 	capabilityBase
 }
 
+// Deprecated
 type RoutingCapability struct {
 	capabilityBase
 }
 
-type DataIngestCapability struct {
+type MultiCapability struct {
 	capabilityBase
 }
 
-// func (c *capabilityBase) setTlsConfig(agSpec *dynatracev1beta1.ActiveGateSpec) {
-// 	if agSpec == nil {
-// 		return
-// 	}
-
-// 	if agSpec.TlsSecretName != "" {
-// 		c.volumes = append(c.volumes,
-// 			corev1.Volume{
-// 				Name: jettyCerts,
-// 				VolumeSource: corev1.VolumeSource{
-// 					Secret: &corev1.SecretVolumeSource{
-// 						SecretName: agSpec.TlsSecretName,
-// 					},
-// 				},
-// 			})
-// 		c.containerVolumeMounts = append(c.containerVolumeMounts,
-// 			corev1.VolumeMount{
-// 				ReadOnly:  true,
-// 				Name:      jettyCerts,
-// 				MountPath: filepath.Join(secretsRootDir, "tls"),
-// 			})
-// 	}
-// }
-
-func NewKubeMonCapability(crProperties *dynatracev1beta1.CapabilityProperties) *KubeMonCapability {
-	c := &KubeMonCapability{
-		capabilityBase{
-			moduleName:     "kubemon",
-			capabilityName: "kubernetes_monitoring",
-			properties:     crProperties,
-			Configuration: Configuration{
-				ServiceAccountOwner: "kubernetes-monitoring",
-			},
-			initContainersTemplates: []corev1.Container{
-				{
-					Name:            initContainerTemplateName,
-					ImagePullPolicy: corev1.PullAlways,
-					WorkingDir:      k8scrt2jksWorkingDir,
-					Command:         []string{"/bin/bash"},
-					Args:            []string{"-c", k8scrt2jksPath},
-					VolumeMounts: []corev1.VolumeMount{
-						{
-							ReadOnly:  false,
-							Name:      trustStoreVolume,
-							MountPath: activeGateSslPath,
-						},
+func (c *capabilityBase) setTlsConfig(agSpec *dynatracev1beta1.ActiveGateSpec) {
+	if agSpec == nil {
+		return
+	}
+
+	if agSpec.TlsSecretName != "" {
+		c.volumes = append(c.volumes,
+			corev1.Volume{
+				Name: jettyCerts,
+				VolumeSource: corev1.VolumeSource{
+					Secret: &corev1.SecretVolumeSource{
+						SecretName: agSpec.TlsSecretName,
 					},
 				},
-			},
-			containerVolumeMounts: []corev1.VolumeMount{{
+			})
+		c.containerVolumeMounts = append(c.containerVolumeMounts,
+			corev1.VolumeMount{
 				ReadOnly:  true,
-				Name:      trustStoreVolume,
-				MountPath: activeGateCacertsPath,
-				SubPath:   k8sCertificateFile,
-			}},
-			volumes: []corev1.Volume{{
-				Name: trustStoreVolume,
-				VolumeSource: corev1.VolumeSource{
-					EmptyDir: &corev1.EmptyDirVolumeSource{},
-				}},
-			},
+				Name:      jettyCerts,
+				MountPath: filepath.Join(secretsRootDir, "tls"),
+			})
+	}
+}
+
+func NewMultiCapability(dk *dynatracev1beta1.DynaKube) *MultiCapability {
+	mc := MultiCapability{
+		capabilityBase{
+			shortName: multiActiveGatePodName,
 		},
 	}
+	if dk == nil || !dk.ActiveGateMode() {
+		return &mc
+	}
+	mc.enabled = true
+	mc.properties = &dk.Spec.ActiveGate.CapabilityProperties
+	capabilityNames := []string{}
+	for _, capName := range dk.Spec.ActiveGate.Capabilities {
+		cap := activeGateCapabilities[capName]()
```

Comment:
There is no protection against invalid DisplayName in [capName]. Operator dies:
```pod/dynatrace-operator-7b66c9cdf9-8czlm   0/1     CrashLoopBackOff   5 (61s ago)   9m5s```

```E1001 11:21:16.848245       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)```

I added 'dummy' capability using 'kubectl edit mydynakube' command:

```
spec:
  activeGate:
    capabilities:
    - kubernetes-monitoring
    - dummy
```

Reply:
This will be fixed with an updated validation webhook

Reply:
yes, the validation webhook will catch this, but I also added a check so in case we want to support this on openshift 3.11 (that doesn't have webhooks) we can do so